### ISSUE TITLE
Indicator Function

### ISSUE THREAD

**Issue by annatelegina** [skmhrk1209/Single-Path-NAS-PyTorch#4]
Hello everyone! @skmhrk1209  @Catosine
Could you please explain us, why did you write the Indicator function for single-path-nas as this: indicator = (norm > threshold).float() - torch.sigmoid(norm - threshold).detach() + torch.sigmoid(norm - threshold) ?
Why can't we use only torch.sigmoid(norm - threshold) in this case ?

**Comment by Catosine** [comment]
Hello @annatelegina ! 

Thanks for your question! Here is the thing: for the indicator function, the expected output should be binary. That is to say, either 0 or 1. However, torch.sigmoid(norm-threshold) is not guaranteed to be only 0 or 1, for it returns a continuous output. Luckily, (norm > threshold).float() returns only 0 or 1, which fulfill our expectation.

As for why the indicator is implemented as:
`indicator = (norm > threshold).float() - torch.sigmoid(norm - threshold).detach() + torch.sigmoid(norm - threshold)`

That is because (norm > threshold).float() is not differentiable. So @skmhrk1209 uses torch.sigmoid(norm - threshold) twice to allow back propagation.

GL,

**Comment by skmhrk1209** [comment]
Hi. Just as @Catosine said, I designed the function that performs as float(norm > threshold) in the forward pass, and sigmoid(norm - threshold) in the backward pass following the paper.

**Comment by annatelegina** [comment]
@Catosine Thank you very much for you answer!
Are you Huawei researcher? I saw the information on your account there. If yes, I would like to discuss some details of these approach. Have you espace account?

**Comment by Catosine** [comment]
@annatelegina yes! my id is f50001770

**Comment by unikcc** [comment]
> Hello @annatelegina !
> 
> Thanks for your question! Here is the thing: for the indicator function, the expected output should be binary. That is to say, either 0 or 1. However, torch.sigmoid(norm-threshold) is not guaranteed to be only 0 or 1, for it returns a continuous output. Luckily, (norm > threshold).float() returns only 0 or 1, which fulfill our expectation.
> 
> As for why the indicator is implemented as: `indicator = (norm > threshold).float() - torch.sigmoid(norm - threshold).detach() + torch.sigmoid(norm - threshold)`
> 
> That is because (norm > threshold).float() is not differentiable. So @skmhrk1209 uses torch.sigmoid(norm - threshold) twice to allow back propagation.
> 

> GL,

Hello @skmhrk1209  and  @Catosine , thanks for your kind reply. But I am still confused about why using `torch.sigmoid(norm - threshold)` twice can make `(norm > threshold).float()` differentiable? Can you supply a more detailed explain? Thank you!
